go-controller/pkg/ovn/controller/apbroute/external_controller.go

```diff
@@ -140,16 +140,59 @@ func newExternalPolicyManager(
 	return &m
 }
 
+// getRoutePolicyFromCache retrieves the cached value of the policy if it exists in the cache, as well as locking the key in case it exists.
 func (m *externalPolicyManager) getRoutePolicyFromCache(policyName string) (*adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute, bool) {
-	return m.routePolicySyncCache.Load(policyName)
+	var (
+		policy *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute
+		found  bool
+	)
+	m.routePolicySyncCache.DoWithLock(policyName, func(policyName string) error {
+		policy, found = m.routePolicySyncCache.Load(policyName)
+		return nil
+	})
+	return policy, found
 }
 
-func (m *externalPolicyManager) updateRoutePolicyInCache(policy *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) {
-	m.routePolicySyncCache.LoadOrStore(policy.Name, policy)
+func (m *externalPolicyManager) storeRoutePolicyInCache(policyInfo *adminpolicybasedrouteapi.AdminPolicyBasedExternalRoute) error {
+	return m.routePolicySyncCache.DoWithLock(policyInfo.Name, func(policyName string) error {
+		m.routePolicySyncCache.Store(policyName, policyInfo)
```

Comment:
@trozet I had to enhance the syncMap an add a `Store` function to make sure I was able to overwrite the stored value. The available `LoadOrStore()` function only allowed me to write once the value and any subsequent calls would return the existing value.
Since the route policy object can change over time, I am forced to update the value of the cache as changes are detected, hence the `Store()` function.
Let me know what you think.

Reply:
Correct. Because I update the stored object. If I had a struct that wrapped the object, I would not need this as I would only be updating a field within the already contained struct. But since I'm changing the reference to the object, I have to rewrite what's there.
Note that I don't have this problem with the namespace info cache, as I only update the fields of the struct.